### ISSUE TITLE
[data grid] Add ability to set minimum / maximum resizable column width

### ISSUE THREAD

**Issue by bryyytl** [mui/mui-x#1561]
Following up after https://github.com/mui-org/material-ui-x/issues/347 and https://github.com/mui-org/material-ui-x/pull/566, I think it would be great to be able to set the minimum / maximum width of a column similarly to how the GridColDef allows the default width to be set.

- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Summary 💡
I could see two different ways this feature could be supported.

1. The minimum / maximum width could be set to apply for all columns (e.g. 100px min-width instead of 50px).
2. The minimum / maximum width could be applied individually per column (similarly to how the default width can be specified).

## Motivation 🔦
Right now, I am trying to prevent users from being able to shrink the columns to the extent where the column headerName no longer appears. The following screenshot shows the two resizable columns being collapsed to the extent that the user cannot see the sorting icon.

![Data Grid - Columns - Material-UI - Google Chrome 5_3_2021 6_09_49 PM](https://user-images.githubusercontent.com/78176335/116940204-9b17ac80-ac3b-11eb-9486-079d7c737eb2.png)

https://material-ui.com/components/data-grid/columns/#resizing

**Comment by fnowacki** [comment]
Also for the `width`/`flex` option, it would be a nice feature!

For now, for large screens, the `flex` option works great to scale columns for the whole width. But for mobile devices... It's totally unreadable :/

Take a look :eyes:

Larger screen (like desktop):
<img width="640" alt="image" src="https://user-images.githubusercontent.com/20989752/117161311-17d49300-adc2-11eb-863a-449efbc3b0c9.png">

Mobile screen (iPhone X):
<img width="250" alt="image" src="https://user-images.githubusercontent.com/20989752/117161411-29b63600-adc2-11eb-83ac-51e5516a59cb.png">

My proposition is, to add a "minWidth" flag 🙌🏻 It should works perfect with `flex` :tada:

**Comment by oliviertassinari** [comment]
@fnowacki your use case is different from the author. Still, I do agree that with the default width, it makes a lot of sense to reproduce the CSS layout model (but a subset that is compatible with virtualization).

AG Grid [has](https://ag-grid.com/react-grid/column-sizing/#column-flex) the same feature for both flexbox layout and resizing, but it's not obvious if it makes sense to merge the two concepts. One could argue that the default width shouldn't impact the capability of a user to adjust the width as he see fits.

**Comment by arheinjohnson** [comment]
In the latest release, is it not possible to resize columns once a default width is set?

**Comment by oliviertassinari** [comment]
We now have #1850 for the flexbox aspect of the min width raised in this PR. For #1561 I would propose we close as "won't fix".

**Comment by fnowacki** [comment]
@oliviertassinari Cool :+1: Many thanks!

**Comment by oliviertassinari** [comment]
With #2101, we have made the `minWidth` property of the column definition also impact how the columns can be resized. We didn't implement the same for the `maxWidth` property.

**Comment by flaviendelangle** [comment]
It is a duplicates of #2174, I am closing this one :+1:

**Comment by oliviertassinari** [comment]
Actually, I'm reopening, this one is specifically about the resize feature, while #2174 is about the default layout. It's close, but it's not clear if the two should be the same.

**Comment by tlmader** [comment]
Also interested in `minWidth` / `maxWidth` properties that do not impact resizing. `minFlexWidth` / `maxFlexWidth` maybe?

**Comment by flaviendelangle** [comment]
I would go the other way around an create `minResizableSize` / `maxResizableSize` (with fallback to `minWidth` / `maxWidth` if not defined)
But open to discussion

**Comment by tlmader** [comment]
That would introduce fewer breaking changes.

How would this work if I wanted to allow full resizing, while setting a minimum flex width? I imagine like this:
```
  {
    field: 'projectId',
    headerName: 'Project',
    minResizableSize: 0,
    minWidth: 200,
    flex: 1,
    renderCell: renderCellText,
  },
```

**Comment by flaviendelangle** [comment]
Right now both flex process and resize process uses `minWidth` / `maxWidth`
If we want to differentiate, we need to create new properties for one of them, with fallback to `minWidth` / `maxWidth` if not defined (to avoir any BC).
The fallback may go away in v6 (not sure what's best here).
Not sure why putting the new properties to the resize or to the flex process differ in the amount of BC.

In your example, for me that grid would take 200px at init (or more if more flex space available).
And it would be resizable below 200px (down to 0px).

**Comment by tlmader** [comment]
You're right, both approaches do not introduce BC. I was in the middle of editing my comment 😄 

Either one would work for me. `minFlexWidth` / `maxFlexWidth` would allow me to do this with one fewer property though:
```
  {
    field: 'projectId',
    headerName: 'Project',
    minFlexWidth: 200,
    flex: 1,
    renderCell: renderCellText,
  },
  ```

**Comment by flaviendelangle** [comment]
:smile: Too fast

Yes I guess the best argument to do one or the other is : which one do we most often want to define without the other ?

**Comment by tlmader** [comment]
Another thought: if MUI ever decided to support other fluid column widths (maybe percentages?), then `minResizableSize` / `maxResizableSize` could be more future-proof.

**Comment by michelengelen** [comment]
This shipped a long time ago, so let us close it. `colDef.minWidth` limits resizing since #2101. `colDef.maxWidth` was added by #3550 and released in [v5.3.0](https://github.com/mui/mui-x/releases/tag/v5.3.0). It limits resizing as well. Both limits also apply to [autosizing](https://mui.com/x/react-data-grid/column-dimensions/#autosizing): "In all the cases, the `colDef.minWidth` and `colDef.maxWidth` options will be respected."

The idea of separate resize-only limits (`minResizableSize` / `maxResizableSize`) from the 2022 discussion never gathered demand. If you still need it, please open a new feature request with your use case.
